### PR TITLE
Update URL-tools for new URL handler

### PR DESCRIPTION
**Pull request by gdude2002**

### CODE REVIEW

File: URL-tools/README.md [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+GitHub
+------
+
+The GitHub API only refers to milestones by ID - the milestone names given in milestone URLs are not supported.
+As such, you will still be able to use these URLs, but you'll have to replace the name with the milestone ID - starting
+with `1` for the earliest milestone. This is less than ideal, but I'm told that they're working on it.
+
+The GitHub handler is unable to handle certain types of URLs. The following URLs and any others under them
+will return GitHub Zen instead, unless Zen has been disabled.
```
Comment (rakiru) [inline URL-tools/README.md]: Zen is disabled by default and falls back to regular handler.

File: URL-tools/plugins/urltools/__init__.py [hunk 2]
```diff
@@ -42,121 +40,47 @@ class URLToolsPlugin(plugin.PluginObject):
 
     YOUTUBE_DESCRIPTION_LENGTH = 75
```
Comment (rakiru) [inline URL-tools/plugins/urltools/__init__.py]: This shit doesn't need to be here.

File: URL-tools/README.md (hunk 1 above)
Comment (gdude2002) [inline URL-tools/README.md]: - [x]

File: URL-tools/plugins/urltools/__init__.py [hunk 3]
```diff
@@ -42,121 +40,47 @@ class URLToolsPlugin(plugin.PluginObject):
...
     @property
     def urls(self):
+        """
+        :rtype: plugins.urls.URLsPlugin
+        """
+
         return self.plugman.get_plugin("URLs")
 
     def setup(self):
         self.storage = StorageManager()
```
Comment (rakiru) [inline URL-tools/plugins/urltools/__init__.py]: This already exists in the superclass.

File: URL-tools/plugins/urltools/__init__.py [hunk 4]
```diff
@@ -42,121 +40,47 @@ class URLToolsPlugin(plugin.PluginObject):
...
+        }
+
+        self.shorteners = {
+            "is.gd": is_gd.IsGdShortener,
+            "nazr.in": nazrin.NazrinShortener,
+            "v.gd": v_gd.VGdShortener,
+            "waa.ai": waa_ai.WaaAiShortener
+        }
 
         self.plugman = PluginManager()
```
Comment (rakiru) [inline URL-tools/plugins/urltools/__init__.py]: This already exists in the superclass.

File: URL-tools/plugins/urltools/__init__.py (hunk 2 above)
Comment (gdude2002) [inline URL-tools/plugins/urltools/__init__.py]: - [x] Done

File: URL-tools/plugins/urltools/__init__.py (hunk 3 above)
Comment (gdude2002) [inline URL-tools/plugins/urltools/__init__.py]: - [x] Done

File: URL-tools/plugins/urltools/__init__.py (hunk 4 above)
Comment (gdude2002) [inline URL-tools/plugins/urltools/__init__.py]: - [x] Done

**Comment by rakiru** [comment]
`RANDOM_SAMPLE_SIZE` in github.py doesn't seem to be used?

**Comment by rakiru** [comment]
I am not reviewing github.py; it's a massive pile of inlineCallbacks.

**Comment by rakiru** [comment]
Youtube ain't done yet, but the old one doesn't work anyway, so it's no loss to go ahead without it.

**Comment by gdude2002** [comment]
> `RANDOM_SAMPLE_SIZE` in github.py doesn't seem to be used?

- [x] Done

> I am not reviewing github.py; it's a massive pile of inlineCallbacks.
> Youtube ain't done yet, but the old one doesn't work anyway, so it's no loss to go ahead without it.

- [x] Done..? xD

File: URL-tools/plugins/urltools/handlers/osu/osu.py [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+URL_BEATMAPS = URL_BASE + "/get_beatmaps"
+URL_MATCH = URL_BASE + "/get_match"
+URL_REPLAY = URL_BASE + "/get_replay"
+URL_SCORES = URL_BASE + "/get_scores"
+URL_USER = URL_BASE + "/get_user"
+URL_USER_BEST = URL_BASE + "/get_user_best"
+URL_USER_RECENT = URL_BASE + "/get_user_recent"
+
+# Attempt to guess the locale
+locale.setlocale(locale.LC_ALL, "")
```
Comment (rakiru) [inline URL-tools/plugins/urltools/handlers/osu/osu.py]: This should really be called once on bot startup then never again. Also, it's called in main urltools init file anyway.

File: URL-tools/plugins/urltools/__init__.py [hunk 6]
```diff
@@ -15,20 +7,26 @@
...
+from plugins.urltools.shorteners import waa_ai
 
-    api_details = {}
-    sites = {}
-    shorteners = {}
+__author__ = 'Gareth Coles'
 
-    plugman = None
+# Attempt to guess the locale.
+locale.setlocale(locale.LC_ALL, "")
```
Comment (rakiru) [inline URL-tools/plugins/urltools/__init__.py]: This should be done in the main bot on startup then never again.

File: URL-tools/plugins/urltools/handlers/osu/osu.py (hunk 5 above)
Comment (gdude2002) [inline URL-tools/plugins/urltools/handlers/osu/osu.py]: - [x] Done, added to run.py instead

File: URL-tools/plugins/urltools/__init__.py (hunk 6 above)
Comment (gdude2002) [inline URL-tools/plugins/urltools/__init__.py]: - [x] Done, added to run.py instead

File: URL-tools/plugins/urltools/handlers/osu/osu.py [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+                score["enabled_mods"] = ", ".join(
+                    get_mods(int(score["enabled_mods"]))
+                )
+        except Exception:
+            pass
+
+        data = beatmap
+
+        if beatmap["approved"] in [
+            "Pending", "WIP", "Graveyard", u"Unknown approval"
```
Comment (rakiru) [inline URL-tools/plugins/urltools/handlers/osu/osu.py]: wat

Why is one of these unicode but not the others?

Comment (gdude2002) [inline URL-tools/plugins/urltools/handlers/osu/osu.py]: - [x] Should've all be Unicode

File: URL-tools/plugins/urltools/handlers/osu/osu.py [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+                message = yield self.beatmap(url, target[1])
+
+        except Exception:
+            self.plugin.logger.exception("Error handling URL: {}".format(url))
+            returnValue(True)
+
+        # At this point, if `message` isn't set then we don't understand the
+        # url, and so we'll just allow it to pass down to the other handlers
+
+        if message and isinstance(message, basestring):
```
Comment (rakiru) [inline URL-tools/plugins/urltools/handlers/osu/osu.py]: Why would it be anything other than a string?

Comment (gdude2002) [inline URL-tools/plugins/urltools/handlers/osu/osu.py]: WHEN IT'S A JAR!!1!111

brb, checking

Comment (gdude2002) [inline URL-tools/plugins/urltools/handlers/osu/osu.py]: - [x] It won't be

File: packages.yml [hunk 9]
```diff
@@ -133,10 +133,12 @@ UrbanDictionary:
         modules: []
         packages: []
 URL-tools:
-    version: 0.2.1
+    version: 1.0.0
     description: "Provides a set of tools for working with the URLs plugin."
     requires:
-        modules: []
+        modules:
+        - requests
```
Comment (rakiru) [inline packages.yml]: txrequests requires requests, so surely this is redundant?

Comment (gdude2002) [inline packages.yml]: - [x] Done, as well as correcting the error in package.yml

**Pull request merged**